src/core/types.h

```diff
@@ -217,31 +219,61 @@ static_assert(sizeof(ArgPassSpec) <= sizeof(void*), "ArgPassSpec doesn't fit in
 static_assert(sizeof(ArgPassSpec) == sizeof(uint32_t), "ArgPassSpec::asInt needs to be updated");
 
 struct ParamNames {
-    bool takes_param_names;
-    std::vector<llvm::StringRef> args;
-    llvm::StringRef vararg, kwarg;
+    // the arguments are either an array of char* or AST_Name* depending on if all_args_contains_names is set or not
+    union NameOrStr {
+        NameOrStr(const char* str) : str(str) {}
+        NameOrStr(AST_Name* name) : name(name) {}
 
-    // This members are only set if the InternedStringPool& constructor is used (aka. source is available)!
-    // They are used as an optimization while interpreting because the AST_Names nodes cache important stuff
-    // (InternedString, lookup_type) which would otherwise have to get recomputed all the time.
-    std::vector<AST_Name*> arg_names;
-    AST_Name* vararg_name, *kwarg_name;
+        const char* str;
+        AST_Name* name;
+    };
+    std::vector<NameOrStr> all_args;
```

Comment:
It's not the biggest win, but one thing we could do is create a non-growable container to use instead of a fixed-size std::vector.  I guess the savings aren't that huge (just a pointer per vector) but sometimes it seems silly to have a std::vector when we don't use any of the capabilities.